[1027][26188] Implement CSP with secure_headers gem

oliverguenther opened the pull request:
**TODOS**

- [x] Remove inline `onclick`, 
- [x] Make link_to_function csp-ready
- [x] Check href="javascript:"
- [x] Rework adminstration user management, uses eval'ed script responses.
  - [x] Removed users/groups JS management
- [x] Check overridden functionality in remote forms of dt customizing (it only decorates the form submission, has nothing to do with remote form handling).
- [x] Rework project settings tabs to avoid js.erb responses (repository)
  - [x] Remove JS tab behavior
- [x] Plugins
  - [x] onclick
  - [x] href="javascript:"

https://community.openproject.com/wp/26188
https://community.openproject.com/wp/1027

oliverguenther committed:
Implement CSP with secure_headers gem

braiam commented on app/views/custom_styles/_inline_css.erb:
```diff
@@ -27,7 +27,7 @@ See doc/COPYRIGHT.rdoc for more details.
 ++#%>
 
-<style>
+<% nonced_javascript_tag do %>
```
Shouldn't this be a `hashed_style_tag` instead? [This seems to suggest](https://github.com/twitter/secureheaders/blob/master/docs/hashes.md) that it should be.

oliverguenther replied:
Not sure what of the article you linked suggests this.

The content is dynamic, thus a statically produced hash (upon asset compilation) would never match. The nonce is generated per-request instead.

braiam commented on app/views/my/page_layout.html.erb:
```diff
@@ -27,6 +27,7 @@ See doc/COPYRIGHT.rdoc for more details.
 
 ++#%>
 <script language="JavaScript">
```
I believe you need to remove the previous script line.

oliverguenther replied:
Good catch, thanks!

oliverguenther committed:
Remove empty javascript: in hrefs

oliverguenther committed:
Allow unsafe-eval for Angular JIT

oliverguenther committed:
Replace toggleFieldset(this)

oliverguenther committed:
Make link_to_function pass CSP

This is a hack to avoid rewriting all these functions, but I've added a
deprecation call

oliverguenther committed:
Fix tab handling in CSP

oliverguenther committed:
Remove onclick with csp-onclick helper

It adds to a nonced javascript tag per page. This allows a faster
migration without changing too much code.

oliverguenther committed:
Rewrite broken password_complexity cuke

oliverguenther committed:
Don't use csp_onclick in rails cell

We don't have the action_view context present, so we can't use `content_for`.

oliverguenther committed:
Fix most of the specs

does not yet address the CSP problem of user memberships

oliverguenther committed:
Replace settings and boards remotes with actual responses

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into housekeeping/csp

oliverguenther committed:
Fix more project settings related specs

oliverguenther committed:
Remove capybara-ng

oliverguenther committed:
Allow sameorigin iframe for help modal

We may want to remove that

oliverguenther committed:
Update secure_headers

oliverguenther committed:
Fix specs changed in plugins

wielinde commented on config/initializers/secure_headers.rb:
```diff
@@ -0,0 +1,42 @@
+    # for now require unsafe-eval for Angular JIT
+    script_src: %w('self' 'unsafe-eval'),
+    # Allow unsafe-inline styles
+    style_src: assets_src + %w('unsafe-inline'),
```
@oliverguenther Request for clarification: Isn't the purpose of all this SCP effort to disallow inline styles and JS? Why is it that we trust inline-styles?

Here https://w3c.github.io/webappsec-csp/#csp-directives it says:
> In either case, developers SHOULD NOT include either 'unsafe-inline', or data: as valid sources in their policies. Both enable XSS attacks by allowing code to be included directly in the document itself; they are best avoided completely.

oliverguenther replied:
Yeah,  ideally, we can disable inline css as well. But that will kill all `<div style="display: none">..` and other things we have lying around. This is just not feasible at the moment and will blow up the PR by several order of magnitude.

That means that in theory, the cross-site styling attack vector is still open (e.g., a [CSS keylogger](https://css-tricks.com/css-keylogger/) would be possible.

We should try and use `hidden` or semantic classes to set initial styles when rendering from the backend. It's definitely possible, but hard.

The same goes for the `unsafe-eval` used for ng2 JIT. We can disable this as soon as we use the AOT build to compile all templates beforehand.

Reference:
https://stackoverflow.com/a/31759553
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src

braiam replied:
@wielinde Sadly, using Angular JIT needs unsave-eval, unlike AoT. https://github.com/angular/angular/issues/12958

braiam replied:
@oliverguenther are there plans to switch to AoT?

oliverguenther replied:
@braiam Yes, we are going to switch before the angular upgrade goes into production (OpenProject 8.0.). There's been difficulties in bringing AOT and ng1 together so we'll need to see what and when that can happen.

wielinde commented on app/views/queries/_filters.html.erb:
```diff
@@ -160,8 +160,9 @@ See docs/COPYRIGHT.rdoc for more details.
           <% end %>
         </div>
         <script type="text/javascript">
```
I believe this `<script...` line to be obsolete.

wielinde commented on spec/views/layouts/base.html.erb_spec.rb:
```diff
@@ -198,7 +198,6 @@
         render
         expect(DesignColor.overwritten.size).to eq(1)
         expect(response).to render_template partial: 'custom_styles/_inline_css'
-        expect(rendered).to match /--primary-color:\s*#{primary_color.get_hexcode}/
```
If we remove that line there will be no spec left, testing that custom styles really make it into the layout. In other words, this line has some feature spec character.

wielinde commented on app/assets/javascripts/specific/toggable_fieldset.js:
```diff
@@ -75,14 +75,21 @@ function toggleFieldset(el) {
+
+      setFieldsetToggleState(fieldset);
+    });
 });
```
I havn't started debugging, but something is broken here. After collapsing a fieldset it opens straigt away:
![bug-fieldset-collapse](https://user-images.githubusercontent.com/327272/37462208-68a1a88a-2851-11e8-90a0-ff80d6a27545.gif)

oliverguenther replied:
Fixed, was caused by the event bubbling up to an upper fieldset. Nested fieldsets.. gwah!

oliverguenther committed:
Fix gemfile

wielinde commented on app/views/common/_tabs.html.erb:
```diff
@@ -40,24 +40,17 @@ See docs/COPYRIGHT.rdoc for more details.
-    <button class="tab-left icon-context icon-arrow-left4" onclick="moveTabLeft(this);"></button>
-    <button class="tab-right icon-context icon-arrow-right5" onclick="moveTabRight(this);"></button>
+    <button class="tab-left icon-context icon-arrow-left4"></button>
+    <button class="tab-right icon-context icon-arrow-right5"></button>
```
On my machine the left/right buttons for tabs don't seem to work anymore. When I change back to `dev` branch, they work again.

![bug-tab-left-right](https://user-images.githubusercontent.com/327272/37475070-21258f14-2872-11e8-86fb-f6cdc090c15c.gif)

oliverguenther replied:
Fixed, was incorrectly assuming a passed event to be a jquery element

wielinde commented on app/controllers/project_settings_controller.rb:
```diff
@@ -0,0 +1,99 @@
+#-- encoding: UTF-8
```
I am not sure where to put this, so I drop it here. When changing the controller name, like you did here, you need to check for implicit i18n consequences. The "no_results_box" is one of those. This text does not get shown properly anymore for instance:         https://github.com/opf/openproject/blob/aa62568debd71a87697ab6f93422a7192091636c/config/locales/en.yml#L143

wielinde reviewed (changes requested):
Hey Oliver, here the first findings.

wielinde committed:
indentions

oliverguenther committed:
Fix superfluous script tag

oliverguenther committed:
Fix duplicate toggling of a fieldset

oliverguenther committed:
Fix move tab left/right functionality

oliverguenther committed:
Fix no_results boxes without breaking existing strings

Renaming to ctrl would have been easier, but that breaks all existing
translations.

oliverguenther committed:
Fix expected alert error class

oliverguenther committed:
Re-add custom style expectation

oliverguenther commented:
@wielinde I believe I have addressed all your remarks. Regarding the renamed no_results, I've added the explicit translations to avoid other languages requiring to update their translation strings.

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into housekeeping/csp

oliverguenther committed:
Allow crowdin in-context with CSP, but add opt-out

Since the crowdin iframe adds all sorts of assets, we will want to
disabled that.

wielinde commented:
LGTM. Thank you.
